pb/generate.sh

```diff
@@ -1,74 +1,197 @@
+#    echo "  protoc-gen-go-grpc --version"
+#    echo ""
+#    echo "Should print 'protoc-gen-go-grpc 1.2.0' (or another more recent version)"
+#    exit 1
+#  fi
+#}
+#
+#main "$@"
+
+#!/bin/bash -u
+# Copyright 2019 dfuse Platform Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
 
 ROOT="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && cd .. && pwd )"
 
 # Protobuf definitions
-PROTO=${PROTO:-"$ROOT/../proto"}
+PROTO=${1:-"$ROOT/proto"}
 
 function main() {
   checks
 
-  current_dir="`pwd`"
-  trap "cd \"$current_dir\"" EXIT
-  pushd "$ROOT/pb" &> /dev/null
+  set -e
+  pushd "$ROOT" >/dev/null
+
+  generate_proto
+
+  popd >/dev/null
 
-  generate "sf/metering/v1/metering.proto"
+#  echo "generate.sh - `date` - `whoami`" > $ROOT/pb/last_generate.txt
+#  echo "dmetering revision: `GIT_DIR=$ROOT/.git git rev-parse HEAD`" >> $ROOT/pb/last_generate.txt
 
   echo "generate.sh - `date` - `whoami`" > $ROOT/pb/last_generate.txt
-  echo "streamingfast/proto revision: `GIT_DIR=$PROTO/.git git rev-parse HEAD`" >> $ROOT/pb/last_generate.txt
+  echo "streamingfast/proto revision: `GIT_DIR=$ROOT/.git git rev-parse HEAD`" >> $ROOT/pb/last_generate.txt
+
+  echo "Done"
 }
 
-# usage:
-# - generate <protoPath>
-# - generate <protoBasePath/> [<file.proto> ...]
-function generate() {
-    base=""
-    if [[ "$#" -gt 1 ]]; then
-      base="$1"; shift
-    fi
-
-    for file in "$@"; do
-      protoc -I$PROTO \
-        --go_out=. --go_opt=paths=source_relative \
-        --go-grpc_out=. --go-grpc_opt=paths=source_relative,require_unimplemented_servers=false \
-         $base$file
-    done
+function generate_proto() {
+  echo "Generating Substreams Protobuf bindings via 'buf'"
+  buf generate proto
 }
 
 function checks() {
+  result=`printf "" | buf --version 2>&1 | grep -Eo '1\.(1[0-9]+|[2-9][0-9]+)\.'`
+  if [[ "$result" == "" ]]; then
+    echo "The 'buf' binary is either missing or is not recent enough (at `which buf || echo N/A`)."
+    echo ""
+    echo "To fix your problem, on Mac OS, perform this command:"
+    echo ""
+    echo "  brew install bufbuild/buf/buf"
+    echo ""
+    echo "On other system, refers to https://docs.buf.build/installation"
+    echo ""
+    echo "If everything is working as expetcted, the command:"
+    echo ""
+    echo "  buf --version"
+    echo ""
+    echo "Should print '1.11.0' (or newer)"
+    exit 1
+  fi
+
   # The old `protoc-gen-go` did not accept any flags. Just using `protoc-gen-go --version` in this
```

Comment:
Using buf remote plugins removes the need for all those checks as we would rely only on `buf`.